app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java

```diff
@@ -0,0 +1,544 @@
+        }
+    }
+
+    private void writeStreamToFile(ByteArrayOutputStream bos, String filename,
+                                   byte[] key) throws IOException {
+        DataOutputStream fileOutputStream = null;
+        try {
+            fileOutputStream = getOutputFileStream(filename, key);
+            bos.writeTo(fileOutputStream);
+        } finally {
+            if (fileOutputStream != null) {
+                fileOutputStream.close();
+            }
+        }
+    }
+
+    protected DataOutputStream getOutputFileStream(String filename,
+                                                   byte[] aesKeyBytes) throws IOException {
+        SecretKeySpec aesKey = new SecretKeySpec(aesKeyBytes, "AES");
+        return new DataOutputStream(EncryptionIO.createFileOutputStream(filename, aesKey));
+    }
+
+    @Override
+    public int add(Externalizable externalizable) {
+        throw new UnsupportedOperationException("Use 'SqlFileBackedStorage.write'");
+    }
+
+    @Override
+    public void update(int id, Externalizable extObj) {
+        SQLiteDatabase db = getDbOrThrow();
+
+        db.beginTransaction();
+        ByteArrayOutputStream bos = null;
+        try {
+            Pair<String, byte[]> filenameAndKey =
+                    HybridFileBackedSqlHelpers.getEntryFilenameAndKey(helper, table, id);
+            String filename = filenameAndKey.first;
+            byte[] fileEncryptionKey = filenameAndKey.second;
+            boolean objectInDb = (filename == null);
+
+            bos = writeExternalizableToStream(extObj);
+            if (blobFitsInDb(bos)) {
+                updateEntryToStoreInDb(extObj, objectInDb, filename, bos, db, id);
+            } else {
+                updateEntryToStoreInFs(extObj, objectInDb, filename,
+                        fileEncryptionKey, bos, db, id);
+            }
+
+            db.setTransactionSuccessful();
+        } catch (IOException e) {
+            e.printStackTrace();
+        } finally {
+            if (bos != null) {
+                try {
+                    bos.close();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+            db.endTransaction();
+        }
+    }
+
+    private void updateEntryToStoreInDb(Externalizable extObj, boolean objectInDb,
+                                        String filename, ByteArrayOutputStream bos,
+                                        SQLiteDatabase db, int id) {
+        ContentValues updatedContentValues =
+                helper.getContentValuesWithCustomData(extObj, bos.toByteArray());
+        if (!objectInDb) {
+            // was stored in file: remove file and store in db
+            updatedContentValues.put(DatabaseHelper.FILE_COL, (String)null);
+            updatedContentValues.put(DatabaseHelper.AES_COL, (byte[])null);
+
+            File dataFile = new File(filename);
+            dataFile.delete();
```

Comment:
If this occurs but the transaction fails our record is now invalid. I would add this to a queue somewhere that manages files to delete so we can't end up in a state where the record becomes corrupted